### DISCUSSION TITLE
Position of Component

### DISCUSSION THREAD

**Discussion by llambdaa** [gyscos/cursive#625] [Q&A] [answered]
Hello there!

I was taking a look at [the position.rs example](https://github.com/gyscos/cursive/blob/main/cursive/examples/position.rs) and was wondering if there is an easy way to get the current position of some component so that I could move it relative to its current location. The view component can be identified using its name or its LayerPosition directly.

From what I could tell by looking at struct StackView, only the position of the topmost element can be retrieved by using offset().
The first hacky approach for moving an arbitrary component I have identified somehow is to use move_layer() to bring it to the front, then use offset(), move the component, and then use move_layer() to bring it back. This way actually works but doesn't feel right :)

So, is there an easy way I have just overlooked?

**Comment by gyscos** [comment, marked as answer]
Hi, and thanks for the question!

"Current position" can be a bit ambiguous:
* The position is often treated as relative to the parent view, but in some cases the "absolute" position (relative to the terminal's origin) can be desired, for example when adding a new layer that should be aligned with the layer underneath (this is what happens when opening popup menus).
* A view can have some negative "virtual offset", when it is wrapped in a `ScrollView` for example. In such a situation, the wrapped view origin could be above and to the left of the terminal, but only a subset of the view would actually be visible.
* Some views can be part of the view tree but not actually printed anywhere. For example, a view hidden in a `HideableView`, or an inactive screen from a `ScreensView`, or [an inactive tab](https://github.com/deinstapel/cursive-tabs).

What matters to you will depend on your use-case.

You say you want to "move [a view] relative to its current location". Is the view currently a layer by itself in a `StackView`? Is it part of a parent layout?

If your view is currently a standalone layer, it's indeed easier to move around without affecting other views. In that case, the absolute offset of the view is what you need. As you noticed, one way to get that is `StackView::offset`, but that only works for the top-level view (that being said, it would make sense to add a `StackView::layer_offset(LayerPosition)`). But there is another way to get this offset.

*EDIT*: I just added this `StackView::layer_offset` method in the latest commit.

One time when the position is made available is when drawing the view. The `Printer` includes various offsets, to determine where in the window to draw the view's content. It's possible to store these offsets, to make them available outside of `View::draw` itself. This is what [`TrackedView`](https://docs.rs/cursive/latest/cursive/views/struct.TrackedView.html) does: it just remembers the last offset used.

This currently ignores any scrolling parent, but you don't care about that if your view is directly a layer from the `StackView`.

**Reply by llambdaa** [reply]
That did the trick! 
Thank you very much for this project (and the commit ;)) - it's helping alot!
